File: WEEK5/bj4_11725.hs
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let lineIndex = 0;

const N = parseInt(input[lineIndex++]);

const visited = new Array(N + 1).fill(false);
const answer = new Array(N + 1).fill(0);
const tree = Array.from({ length: N + 1 }, () => []);

for (let i = 1; i < N; i++) {
    const [n1, n2] = input[lineIndex++].split(' ').map(Number);
    tree[n1].push(n2);
    tree[n2].push(n1);
}

function DFS(number) {
    visited[number] = true;
    for (const child of tree[number]) {
        if (!visited[child]) {
            answer[child] = number;
            DFS(child);
        }
    }
}

DFS(1);

const sb = [];
for (let i = 2; i <= N; i++) {
    sb.push(answer[i]);
}
console.log(sb.join('\n'));

File: WEEK5/bj4_11725.py
import sys

sys.setrecursionlimit(10**6 + 1)
input = sys.stdin.readline

N = int(input())
visited = [False] * (N + 1)
answer = [0] * (N + 1)
tree = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    n1, n2 = map(int, input().split())
    tree[n1].append(n2)
    tree[n2].append(n1)


def DFS(number):
    visited[number] = True
    for child in tree[number]:
        if not visited[child]:
            answer[child] = number
            DFS(child)


DFS(1)

for i in range(2, N + 1):
    print(answer[i])
